compiler/noirc_evaluator/src/ssa/opt/constant_folding.rs

```diff
@@ -188,12 +190,56 @@ pub(crate) struct BrilligInfo<'a> {
     brillig_functions: &'a BTreeMap<FunctionId, Function>,
 }
 
+/// Records a simplified equivalent of an [`Instruction`]s along with the blocks in which the
+/// constraint that advised the simplification has been encountered.
+///
+/// For more information see [`ConstraintSimplificationCache`].
+#[derive(Default)]
+struct SimplificationCache {
+    /// Simplified expressions where we found them.
+    /// Using a `BTreeMap` for deterministic enumeration.
+    simplifications: BTreeMap<BasicBlockId, ValueId>,
+}
+
+impl SimplificationCache {
+    /// Called with a newly encountered simplification.
+    fn add(&mut self, dfg: &DataFlowGraph, simple: ValueId, block: BasicBlockId) {
+        let existing = self.simplifications.entry(block).or_insert(simple);
+        // Keep the simpler expression in this block.
+        if *existing != simple {
+            match simplify(dfg, *existing, simple) {
+                Some((complex, simple)) if *existing == complex => {
+                    *existing = simple;
+                }
+                _ => {}
+            }
+        }
```

Comment:
```suggestion
       self.simplifications
            .entry(block)
            .and_modify(|existing| {
                // `SimplificationCache` may already hold a simplification in this block
                // so we check whether `simple` is a better simplification than the current one.
                if let Some((_, simpler)) = simplify(dfg, *existing, simple) {
                    *existing = simpler;
                };
            })
            .or_insert(simple);
```

I think this better shows the intent of this function


`cargo fmt` is gonna not like this so will need to be run.

Reply:
Nice!